feat(client): add support for JS connectors

jkomyno opened the pull request:
**DO NOT MERGE (yet)**

This PR includes:
- ability to pass an instance of a JS Connector (aka, Node.js Driver) to the Prisma Client
- import of temporary npm libraries extracted from https://github.com/prisma/prisma-engines/pull/4097 and following the [Distribution doc](https://www.notion.so/prismaio/Distribution-499d727a48bf4403a5f8b9a3f84d4989). These libraries are currently scoped under an internal npm domain, `@jkomyno/*`. These libraries are:
  - `@jkomyno/prisma-js-connector-utils` (in the future: `@prisma/js-connector-utils`)
  - `@jkomyno/prisma-planetscale-js-connector` (in the future: `@prisma/planetscale-js-connector`)
  - `@jkomyno/prisma-neon-js-connector` (in the future: `@prisma/neon-js-connector`)
- reproduction (in `./reproductions/js-connectors`) that showcases how to use Prisma Client + JS Connectors

For the time being, we expect internal developers to try out JS Connectors + Prisma Client using the following template:

```typescript
import { PrismaClient } from '@prisma/client'
import { createPlanetScaleConnector } from '@jkomyno/prisma-planetscale-js-connector'

const connectionString = `${process.env.JS_PLANETSCALE_DATABASE_URL as string}`

const jsConnector = createPlanetScaleConnector({
  url: connectionString,
})

const prisma = new PrismaClient({ jsConnector })
```

---

/integration

jkomyno committed:
feat(client): add support for JS Connectors in Prisma Client

jkomyno committed:
feat: add reproduction for js-connectors

jkomyno commented on reproductions/js-connectors/.envrc:
```diff
@@ -0,0 +1,7 @@
+export JS_PLANETSCALE_DATABASE_URL="mysql://rkktvtka0swhgcp6m02u:pscale_pw_nw60nUUQ69OEpdAy4liK19yQdjEU4NUSiL9BqzfG2r2@aws.connect.psdb.cloud/cf-users?sslaccept=strict"
+export JS_PLANETSCALE_SHADOW_DATABASE_URL="mysql://7yjjs4l82xwvfcqig291:pscale_pw_5rB0hFl5mTZnltvCeKY2ZwErhJQEdr8NkZ4ythiS24K@aws.connect.psdb.cloud/cf-users?sslaccept=strict"
+export JS_NEON_DATABASE_URL="postgres://jkomyno:aiPsj5FrEN9t@ep-restless-cake-780415.eu-central-1.aws.neon.tech/neondb"
+export JS_NEON_SHADOW_DATABASE_URL="postgres://jkomyno:aiPsj5FrEN9t@ep-still-rain-49733054.eu-central-1.aws.neon.tech/neondb"
```
F, this is the result of a misaligned .gitignore file between the prisma-engines repo's subfolder and this one. Thanks for spotting. I've received an automatic email by Planetscale, as they've spotted this as well

janpio replied:
Nice, share that email in #db-tidbits maybe, interesting to know about.

jkomyno committed:
chore: removed .envrc

jkomyno committed:
chore: add NEON urls to .envrc.example

jkomyno committed:
fix: panic with PlanetScale on Node.js < v18

jkomyno committed:
chore: comment bigint column for MySQL's type_test

janpio commented on reproductions/js-connectors/README.md:
```diff
@@ -0,0 +1,47 @@
+
+- Set the provider in [./prisma/postgres-neon/schema.prisma](./prisma/postgres-neon/schema.prisma) to `postgres`.
+- Run `npx prisma db push --schema ./prisma/postgres-neon/schema.prisma`
+- Run `npx prisma migrate deploy --schema ./prisma/postgres-neon/schema.prisma`
```
Nit: Technically this is an abuse of `prisma migrate`, with pure `INSERT` this should better be run via `db execute` instead.

jkomyno committed:
fix(js-connectos): Prisma setup for JS Connectors

jkomyno committed:
chore(js-connectors): add command output to README

jkomyno committed:
Merge branch 'main' into feat/js-connectors-in-client

jkomyno committed:
Merge branch 'main' into feat/js-connectors-in-client

jkomyno committed:
chore(js-connectors): update deps

jkomyno committed:
chore(js-connectors): checked support for transactions in Neon

jkomyno committed:
test(js-connectors): add test for explicit batch transactions with isolation levels

jkomyno committed:
Merge branch 'main' into feat/js-connectors-in-client

jkomyno commented:
/integration

jkomyno committed:
Merge branch 'main' into feat/js-connectors-in-client

jkomyno committed:
chore: bump JS connectors' versions to avoid panic

jkomyno committed:
Merge branch 'main' into feat/js-connectors-in-client

jkomyno committed:
feat(js-connectors): simplify smoke tests after https://github.com/prisma/prisma-engines/pull/4113

Jolg42 committed:
Merge branch 'main' into feat/js-connectors-in-client

Jolg42 committed:
Merge branch 'main' into feat/js-connectors-in-client

Jolg42 committed:
Merge branch 'main' into feat/js-connectors-in-client

jkomyno committed:
chore(js-connectors): remove references to shadow URLs

jkomyno committed:
chore(js-connectors): update neon's output in README

jharrell commented on reproductions/js-connectors/.envrc.example:
```diff
@@ -0,0 +1,5 @@
+export JS_PLANETSCALE_DATABASE_URL="mysql://USER:PASSWORD@aws.connect.psdb.cloud/DATABASE?sslaccept=strict"
+export JS_PLANETSCALE_SHADOW_DATABASE_URL="mysql://USER:PASSWORD@aws.connect.psdb.cloud/DATABASE?sslaccept=strict"
+
+export JS_NEON_DATABASE_URL="postgres://USER:PASSWORD@DATABASE-pooler.eu-central-1.aws.neon.tech/neondb?pgbouncer=true&connect_timeout=10"
+export JS_NEON_SHADOW_DATABASE_URL="postgres://USER:PASSWORD@DATABASE-pooler.eu-central-1.aws.neon.tech/neondb?pgbouncer=true&connect_timeout=10"
```
```suggestion
export JS_PLANETSCALE_DATABASE_URL="mysql://USER:PASSWORD@aws.connect.psdb.cloud/DATABASE?sslaccept=strict"

export JS_NEON_DATABASE_URL="postgres://USER:PASSWORD@DATABASE-pooler.eu-central-1.aws.neon.tech/neondb?pgbouncer=true&connect_timeout=10"
```

jkomyno committed:
chore: update .envrc.example

Co-authored-by: Jon Harrell <4829245+jharrell@users.noreply.github.com>

jkomyno committed:
Merge branch 'main' into feat/js-connectors-in-client

jkomyno committed:
chore: merge main, fix conflicts, bump "@jkomyno/prisma-js-connector-utils"'s version, fix types

jkomyno committed:
chore: bump versions, use custom fetch in neon

jkomyno committed:
feat(js-connectors): showcase interactive transactions

jkomyno committed:
Merge branch 'feat/js-connectors-in-client' of github.com:prisma/prisma into feat/js-connectors-in-client

jkomyno committed:
chore(js-connectors): add @jkomyno/prisma-js-connectors-utils to tscPlugin

jkomyno committed:
Merge branch 'main' into feat/js-connectors-in-client

jkomyno committed:
chore(js-connectors): add missing ColumnType

jkomyno committed:
chore(js-connectors): fix ColumnType declarations due to type tests

jkomyno committed:
chore(js-connectors): fix some type tests

jkomyno committed:
chore: merge main, fix conflicts, fix types?

jkomyno committed:
fix: update snapshot

jkomyno committed:
chore: merge main, fix conflicts

Jolg42 commented on packages/client/src/generation/TSClient/PrismaClient.ts:
```diff
@@ -421,12 +421,47 @@ get ${methodName}(): Prisma.${m.model}Delegate<ExtArgs>;`
+  args: Array<unknown>
+}
+
+type Connector = {
+  readonly flavour: 'mysql' | 'postgres'
+  queryRaw: (params: Query) => Promise<ResultSet>
+  executeRaw: (params: Query) => Promise<number>
+  version: () => Promise<string | undefined>
+  isHealthy: () => boolean
+  close: () => Promise<void>
+}
```
TODO, use `$Utils.JsPromise` (added by Serhii recently), check https://github.com/prisma/prisma/blob/69a131b39932d3a41260fc3af4796a1e8b63660c/packages/client/src/runtime/core/types/Utils.ts#L126-L131

jkomyno committed:
chore: use .JsPromise

jkomyno committed:
fix: update library snap

jkomyno committed:
fix: update binary and data-proxy snap

jkomyno committed:
Merge branch 'main' into feat/js-connectors-in-client

jkomyno committed:
chore: fix excessive backtick

jkomyno committed:
chore: fix backtick

jkomyno committed:
Merge branch 'main' into feat/js-connectors-in-client

jkomyno committed:
Merge branch 'main' into feat/js-connectors-in-client

jkomyno commented:
As of c990960b2fcfae07f29dda2aaa9a6fc931c4e0d4, https://github.com/prisma/team-orm/issues/277 can be considered closed.

Jolg42 committed:
Merge remote-tracking branch 'origin' into feat/js-connectors-in-client

jkomyno committed:
Merge branch 'feat/js-connectors-in-client' of github.com:prisma/prisma into feat/js-connectors-in-client

jkomyno committed:
feat(js-connectors): add support for "prisma.$queryRaw", fix https://github.com/prisma/team-orm/issues/308

jkomyno committed:
chore: removed drafts

jkomyno committed:
Merge branch 'main' into feat/js-connectors-in-client

jkomyno committed:
feat(js-connectors): use new transactions

jkomyno committed:
chore: merge main, fix conflicts

jkomyno committed:
chore(js-connectors): update types

jkomyno committed:
chore(js-connectors): replace Promise -> $Utils.JsPromise

jkomyno committed:
chore(js-connectors): fix spacing in snaps

jkomyno committed:
chore(js-connectors): add pg-js-connector

jkomyno committed:
Merge branch 'main' into feat/js-connectors-in-client

jkomyno committed:
chore(js-connectors): fix BigInt serialisation issues by replacing "JSON" with "superjson"

jkomyno committed:
Merge branch 'main' into feat/js-connectors-in-client

jkomyno committed:
Merge branch 'main' into feat/js-connectors-in-client

jkomyno committed:
fix(js-connectors): ColumnType definition

Jolg42 committed:
Merge remote-tracking branch 'origin' into feat/js-connectors-in-client

Jolg42 committed:
Merge branch 'main' into feat/js-connectors-in-client

Jolg42 committed:
Merge commit '7e2ef69880f9e12aa61877e2af2b6693ddc66519' into feat/js-connectors-in-client

* commit '7e2ef69880f9e12aa61877e2af2b6693ddc66519':
  chore(deps): update engines to 5.3.0-16.0a64c821ccc1eab2178ebe0857ccbcb019b869f6 (#20898)
  test(e2e): remove wrangler in unsupported-edge-error (#20886)
  chore(deps): update engines to 5.3.0-15.ed0165d15cf71994f95ac92c3b73c8a54988bf2a (#20897)
  chore(deps): update vitess/vttestserver:mysql80 docker digest to 5b02b2c (#20608)
  fix(deps): update dependency node-fetch to v2.7.0 (#20892)
  chore(deps): update jest to v29.6.4 (#20891)
  chore(deps): update dependency typescript to v5.2.2 (#20879)
  chore(deps): update dependency @types/node to v18.17.12 (#20809)
  chore(deps): update engines to 5.3.0-13.c021fae20e568b13571bf9b1940d8813bff7ba88 (#20888)
  ci: bump e2e timeout in test-template.yml
  fix(deps): update dependency mssql to v9.2.0 (#20880)
  chore: use dotenv-cli for db urls (#20847)
  ci: type check codebase (#20855)
  fix(generate): Empty line before usage instructions (#20859)

# Conflicts:
#	packages/client/package.json
#	pnpm-lock.yaml

Jolg42 committed:
Merge branch 'main' into feat/js-connectors-in-client

Jolg42 committed:
Merge branch 'main' into feat/js-connectors-in-client

SevInf committed:
Integrate latest changes

- Update connector packages
- Revert back to default connectior names
- Handle JS errors in connectors

jkomyno committed:
chore: merge main, fix conflicst

jkomyno committed:
chore: merge main, fix conflicst

jkomyno committed:
chore: update types

jkomyno committed:
chore: update types

jkomyno committed:
feat(js-connectors): align schema setup and tests to prisma-engines

jkomyno committed:
feat(js-connectors): throw Client initialisation error if "jsConnectors" option is passed to BinaryEngine

jkomyno committed:
chore(js-connectors): update type name "Result" into "NapiResult" due to conflicts

SevInf commented on packages/client/src/__tests__/integration/happy/not-so-exhaustive-schema-mongo/__snapshots__/library.test.ts.snap:
```diff
@@ -2284,35 +2284,35 @@ export namespace Prisma {
     * and returning the type-aware result set of the query.
     */
-    queryRaw(params: Query): $Utils.JsPromise<Result<ResultSet>>
+    queryRaw(params: Query): $Utils.JsPromise<NapiResult<ResultSet>>
```
I think we can just use fully-qualified name (with namespace here): `Prisma.Result`

jkomyno committed:
chore(js-connectors): minor snapshot update

jkomyno committed:
chore(js-connectors): minor snapshot fix

jkomyno committed:
chore: merge main, fix conflicts

Jolg42 committed:
Merge branch 'main' into feat/js-connectors-in-client

Jolg42 commented:
The story continues here https://github.com/prisma/prisma/pull/20965
